crates/pathfinder/src/merkle_tree.rs

```diff
@@ -0,0 +1,1089 @@
+        match self.root.borrow().hash() {
+            Some(hash) if hash != ZERO_HASH => self
+                .storage
+                .decrement_ref_count(hash)
+                .context("Failed to delete tree root"),
+            _ => Ok(()),
+        }
+    }
+
+    /// Loads an existing tree or creates a new one if it does not yet exist.
+    ///
+    /// Use the [ZERO_HASH] as root if the tree does not yet exist, will otherwise
+    /// error if the given hash does not exist.
+    ///
+    /// The transaction is used for all storage interactions. The transaction
+    /// should therefore be committed after all tree mutations are completed.
+    ///
+    /// Uses an [RcNodeStorage] as backing storage.
+    ///
+    /// ### Warning
+    ///
+    /// None of the [RcNodeStorage] functions rollback on failure. This means that if any error
+    /// is encountered, the transaction should be rolled back to prevent database corruption.
+    pub fn load(
+        table: String,
+        transaction: &'a Transaction,
+        root: StarkHash,
+    ) -> anyhow::Result<Self> {
+        let storage = RcNodeStorage::open(table, transaction)?;
+
+        // Create a tree with root node unresolved, so we can use the resolve function.
+        // Bit clumsy, but oh well.
+        let root_node = Rc::new(RefCell::new(Node::Unresolved(root)));
+        let mut tree = Self {
+            storage,
+            root: root_node,
+        };
+
+        if root != ZERO_HASH {
+            // Resolve non-zero root node to check that it does exist.
+            let root_node = tree
+                .resolve(root, 0)
+                .context("Failed to resolve root node")?;
+            tree.root = Rc::new(RefCell::new(root_node));
+        }
+
+        Ok(tree)
+    }
+
+    /// Persists all changes to storage and returns the new root hash.
+    ///
+    /// Note that the root is reference counted in storage. Committing the
+    /// same tree again will therefore increment the count again.
+    pub fn commit(self) -> anyhow::Result<StarkHash> {
+        // Go through tree, collect dirty nodes, calculate their hashes and
+        // persist them. Take care to increment ref counts of child nodes. So in order
+        // to do this correctly, will have to start back-to-front.
+        self.commit_subtree(&mut *self.root.borrow_mut())?;
+        // unwrap is safe as `commit_subtree` will set the hash.
+        let root = self.root.borrow().hash().unwrap();
+        self.storage.increment_ref_count(root)?;
+        Ok(root)
+    }
+
+    /// Persists any changes in this subtree to storage.
+    ///
+    /// This necessitates recursively calculating the hash of, and
+    /// in turn persisting, any changed child nodes. This is necessary
+    /// as the parent node's hash relies on its childrens hashes.
+    ///
+    /// In effect, the entire subtree gets persisted.
+    fn commit_subtree(&self, node: &mut Node) -> anyhow::Result<()> {
```

Comment:
Just thinking loudly - maybe I missed something ... but I guess we don't need any limit on recursion here? I can imagine a tree of depth bigger than the max depth of 251 not being asserted anywhere (the height check is in edge node only?) constructed from binary and leaf nodes only.